### PR TITLE
Enable onephase revised

### PR DESCRIPTION
**Pull request by atgeirr**
This is a revised version of #510. Basically it is what I would have requested if reviewing it.

### CODE REVIEW

File: ewoms/models/blackoil/blackoilintensivequantities.hh [hunk 1]
```diff
@@ -335,7 +344,14 @@ public:
         Scalar rockCompressibility = problem.rockCompressibility(elemCtx, dofIdx, timeIdx);
         if (rockCompressibility > 0.0) {
             Scalar rockRefPressure = problem.rockReferencePressure(elemCtx, dofIdx, timeIdx);
-            Evaluation x = rockCompressibility*(fluidState_.pressure(oilPhaseIdx) - rockRefPressure);
+            Evaluation x;
+            if (FluidSystem::phaseIsActive(oilPhaseIdx)) {
+                x = rockCompressibility*(fluidState_.pressure(oilPhaseIdx) - rockRefPressure);
```
Comment (totto82) [inline ewoms/models/blackoil/blackoilintensivequantities.hh]: Maybe we should add a method in the fluidState that return the refereance pressure and temperature.  Just fluidState_.pressure() and fluidState_.temperature(). And then set the reference phase in the BlackoilXXXIndices class. 
The need to access the reference values applies to many places in both the simulator part and the well part of the code.

File: ewoms/models/blackoil/blackoilprimaryvariables.hh [hunk 2]
```diff
@@ -284,7 +289,16 @@ public:
...
 
         // assign the actual primary variables
-        if (primaryVarsMeaning() == Sw_po_Sg) {
+        if (primaryVarsMeaning() == OnePhase_p) {
+            if (waterEnabled) {
+                (*this)[waterSaturationIdx] = FsToolbox::value(fluidState.saturation(waterPhaseIdx));
+                (*this)[pressureSwitchIdx] = FsToolbox::value(fluidState.pressure(waterPhaseIdx));
+            } else {
+                throw std::logic_error("Only pure ware is presently allowed");
+            }
```
Comment (totto82) [inline ewoms/models/blackoil/blackoilprimaryvariables.hh]: Part of the code implements onephase more general then only for water.  Do we have a plan to also support only oil and only gas?

File: ewoms/models/blackoil/blackoilratevector.hh [hunk 3]
```diff
@@ -81,17 +81,6 @@ public:
     BlackOilRateVector(Scalar value) : ParentType(Toolbox::createConstant(value))
     {}
 
-    template <class Eval = Evaluation>
-    BlackOilRateVector(const typename std::enable_if<std::is_same<Eval, Evaluation>::value, Evaluation>::type& value) : ParentType(value)
-    {}
-
```
Comment (totto82) [inline ewoms/models/blackoil/blackoilratevector.hh]: Why are this removed

File: ewoms/models/blackoil/blackoilenergymodules.hh [hunk 4]
```diff
@@ -376,7 +376,6 @@ public:
         // and the thermal condictivity coefficients
         for (int phaseIdx = 0; phaseIdx < numPhases; ++ phaseIdx) {
             if (!FluidSystem::phaseIsActive(phaseIdx)) {
-                fs.setEnthalpy(phaseIdx, 0.0);
                 continue;
```
Comment (totto82) [inline ewoms/models/blackoil/blackoilenergymodules.hh]: Why this change

File: ewoms/models/blackoil/blackoilintensivequantities.hh (hunk 1 above)
Comment (atgeirr) [reply ewoms/models/blackoil/blackoilintensivequantities.hh]: I think that might be a good idea, although we should be careful not to use the wrong pressure. I do not know a simple general rule that tells me which quantities should be evaluated at phase pressures and which should be at "reference" pressure (putting reference in quotes here, as the term is also used for things like rockRefPressure in the above code, which is a different concept...).

File: ewoms/models/blackoil/blackoilprimaryvariables.hh (hunk 2 above)
Comment (atgeirr) [reply ewoms/models/blackoil/blackoilprimaryvariables.hh]: I don't think there are concrete plans now, but it should not be excluded (as you cannot do everything with water, such as table-based PVT).

**Comment by OPMUSER** [comment]
throw std::logic_error("Only pure *ware* is presently allowed");

water

OPM User
opmuser0@gmail.com

On 10-Oct-19 14:36, Atgeirr Flø Rasmussen wrote:
> throw std::logic_error("Only pure ware is presently allowed");



-- 
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

File: ewoms/models/blackoil/blackoilratevector.hh (hunk 3 above)
Comment (atgeirr) [reply ewoms/models/blackoil/blackoilratevector.hh]: This constructor caused errors for single-phase cases, and seems to be unused for other cases.

File: ewoms/models/blackoil/blackoilenergymodules.hh (hunk 4 above)
Comment (atgeirr) [reply ewoms/models/blackoil/blackoilenergymodules.hh]: Temporary reverted, will check in testing.

**Comment by akva2** [comment]
This broke the post-build.

File: ewoms/models/blackoil/blackoilintensivequantities.hh (hunk 1 above)
Comment (totto82) [reply ewoms/models/blackoil/blackoilintensivequantities.hh]: I still think the solution you merged is not very robust. Now the code that picks the right pressure depending on the available phase (or model) is spread around in the code.  In this code the priority is oil water gas. Any reason at all for that particular choice?

Comment (atgeirr) [reply ewoms/models/blackoil/blackoilintensivequantities.hh]: I agree (still) that we should move this logic to the fluid state object, but I did not think it was a good idea to add that refactoring to these PRs (I see now that you did not agree with that...)

The "oil->water->gas" priority order is arbitrary, but not more so than our previous choice "oil (only)". Did you already have an idea or prototype for changing the fluid state? If so, a PR for that would be the natural place to discuss and test this. Instead of a hard-coded order, the "reference" phase could be a (compile-time) property for example.
